zio-http/src/test/scala/zio/http/TestServer.scala

```diff
@@ -34,80 +29,74 @@ final case class TestServer[State](
    *   Response that the Server will send
    *
    * @example
-   *   {{{ ZIO.serviceWithZIO[TestServer[Unit]]( _.addRequestResponse(
-   *   Request.get(url = URL.root.setPort(port = ???)), Response(Status.Ok)
-   *   ).install ) }}}
+   *   {{{
+   *   TestServer.addRequestResponse(Request.get(url = URL.root.setPort(port = ???)), Response(Status.Ok))
+   *   }}}
    *
    * @return
    *   The TestSever with new behavior.
    */
   def addRequestResponse(
     expectedRequest: Request,
     response: Response,
-  ): TestServer[State] = {
-    val handler: PartialFunction[(State, Request), (State, Response)] = {
-      case (state, realRequest) if {
+  ): ZIO[Any, Nothing, Unit] = {
+    val handler: PartialFunction[Request, ZIO[Any, Nothing, Response]] = {
+      case (realRequest) if {
             // The way that the Client breaks apart and re-assembles the request prevents a straightforward
             //    expectedRequest == realRequest
             expectedRequest.url.relative == realRequest.url &&
             expectedRequest.method == realRequest.method &&
             expectedRequest.headers.toSet.forall(expectedHeader => realRequest.headers.toSet.contains(expectedHeader))
           } =>
-        (state, response)
+        ZIO.succeed(response)
     }
-    addHandlerState(handler)
+    addHandler(handler)
   }
 
   /**
-   * Define stateless behavior for the Server.
-   *
+   * Add new behavior to Server
    * @param pf
-   *   The stateless behavior
-   *
+   *   New behavior
    * @return
    *   The TestSever with new behavior.
    *
    * @example
    *   {{{
-   *   ZIO.serviceWithZIO[TestServer[Unit]](_.addHandler { case _: Request =>
-   *       Response(Status.Ok)
-   *   }
+   *  for {
+   *    state <- Ref.make(0)
+   *    testRequest <- requestToCorrectPort
+   *    _           <- TestServer.addHandler{ case (_: Request) =>
+   *      for {
+   *        curState <- state.getAndUpdate(_ + 1)
+   *      } yield {
+   *        if (curState > 0)
+   *          Response(Status.InternalServerError)
+   *        else
+   *          Response(Status.Ok)
+   *      }
+   *    }
    *   }}}
    */
-  def addHandler(pf: PartialFunction[Request, Response]): TestServer[State] = {
-    val handler: PartialFunction[(State, Request), (State, Response)] = {
-      case (state, request) if pf.isDefinedAt(request) => (state, pf(request))
-    }
-    addHandlerState(handler)
+  def addHandler(
```

Comment:
You can make this more polymorphic:

```suggestion
  def addHandler[R, E](
      pf: PartialFunction[Request, ZIO[R, E, Response]],
  ): ZIO[R, E, Unit] = {
```

Reply:
Hmm, well, you can do that for `R`, but not for `E`.

The most polymorphic you can do is:

```suggestion
def addHandler[R](
      pf: PartialFunction[Request, ZIO[R, Throwable, Response]],
  ): ZIO[R, Nothing, Unit] = {
```

assuming we can fail with `Throwable` in the handler.